File: 백준/Gold/7576. 토마토/토마토.py
import os, io
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline

M, N = map(int, input().split())
arr = [[-1] * (M + 2)] + [[-1] + list(map(int, input().split())) + [-1] for _ in range(N)] + [[-1] * (M + 2)]
d = [(1, 0), (-1, 0), (0, 1), (0, -1)]
q = []
ans = -1
cnt = 0

for i in range(1, N + 1):
	for j in range(1, M + 1):
		if arr[i][j] == 0:
			cnt += 1
		elif arr[i][j] == 1:
			q.append((i, j))

while q:
	new_q = []
	for i, j in q:
		for dy, dx in d:
			y, x = i + dy, j + dx
			if arr[y][x] == 0:
				cnt -= 1
				arr[y][x] = 1
				new_q.append((y, x))
	q = new_q
	ans += 1

print(ans if cnt == 0 else -1)

File: 백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
    public static void main(String[] args) throws Exception {
        Solution solution = new Solution();
        System.out.println(solution.solution());
    }
}

class Solution {
    int solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int M = Integer.parseInt(input[0]);
        int N = Integer.parseInt(input[1]);

        int remainingTomatoes = 0;
        Queue<Integer[]> q = new LinkedList<>();
        int[][] grid = new int[N][M];
        for (int i = 0; i < N; i++) {
            input = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                int cur = Integer.parseInt(input[j]);
                if (cur == 1) q.add(new Integer[]{i, j, 0});
                if (cur == 0) remainingTomatoes++;
                grid[i][j] = cur;
            }
        };

        int answer = 0;
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        boolean[][] visited = new boolean[N][M];
        while (!q.isEmpty()) {
            Integer[] cur = q.poll();
            int x = cur[0], y = cur[1], cnt = cur[2];

            if (answer < cnt) answer = cnt;
            for (int d = 0; d < 4; d++) {
                int nx = x + directions[d][0];
                int ny = y + directions[d][1];
                if (canPush(nx, N, ny, M, visited, grid)) {
                    visited[nx][ny] = true;
                    remainingTomatoes--;
                    q.add(new Integer[]{nx, ny, cnt + 1});
                }
            }
        }

        return remainingTomatoes == 0 ? answer : -1;
    }

    private boolean canPush(int nx, int N, int ny, int M, boolean[][] visited, int[][] grid) {
        return 0 <= nx && nx < N && 0 <= ny && ny < M && !visited[nx][ny] && grid[nx][ny] == 0;
    }
}